### PR TITLE
Add built-in deterministic NLP scorers for model evaluation

### LINKED ISSUE #22424: [FR] Built-in deterministic NLP scorers — Readability, BERTScore, METEOR, Sentiment, LevenshteinRatio
<!-- issue-warning -->
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

                                                                                                                                                                                                                                                                        
 ### Willingness to contribute                                                                                                                           
                  
  Yes. I can contribute this feature independently.                                                                                                       
   
  ### Proposal Summary                                                                                                                                    
                  
MLflow v3.4.0 deprecated legacy NLP metrics (`flesch_kincaid_grade_level`, `ari_grade_level`, `rouge1`, `bleu`, etc.) and directed users to the new GenAI scorer framework. ROUGE and BLEU were replaced via the RAGAS third-party integration, but **readability metrics were never migrated**. Users who relied on `mlflow.metrics.flesch_kincaid_grade_level()` have no scorer equivalent today.                                                                
                  
This proposal adds **5 built-in deterministic NLP scorers** to fill this migration gap and extend MLflow's evaluation coverage with standard NLP metrics that are not available in any existing integration:
                                                                                                                                                          
  | Scorer | What it computes | Replaces / Fills |
  |---|---|---|
  | `Readability` | Flesch-Kincaid grade level and Automated Readability Index | **Migration Gap** - `mlflow.metrics.flesch_kincaid_grade_level()` and `mlflow.metrics.ari_grade_level()` deprecated in 3.4.0 with no scorer replacement |                                                                     
  | `BERTScore` | Semantic similarity between output and reference using BERT embeddings | **New** - not in any MLflow integration |
  | `METEOR` | Paraphrase-aware generation quality (standard in NLG evaluation) | **New** - not in any MLflow integration |                               
  | `Sentiment` | VADER sentiment polarity score | **New** - not in any MLflow integration |                                                              
  | `LevenshteinRatio` | Normalized edit distance between output and reference | **New** - not in any MLflow integration |
                                                                                                                                                          
All dependencies are optional — scorers raise a clear error if the dependency is missing.
                                                                                                                                                          
  ### Motivation  

  > #### What is the use case for this feature?

  1. **Migration gap**: Users of the deprecated `flesch_kincaid_grade_level()` and `ari_grade_level()` metrics have no path to the new scorer framework.  
  The migration guide at https://mlflow.org/docs/latest/genai/eval-monitor/legacy-llm-evaluation/ tells them to use scorers, but no readability scorer
  exists.                                                                                                                                                 
                  
  2. **Standard NLP evaluation**: BERTScore and METEOR are standard metrics in NLG evaluation papers and benchmarks (SQuAD, CNN/DailyMail, WMT). They're  
  free (no LLM calls), deterministic, and more semantically meaningful than string overlap metrics like ROUGE.
                                                                                                                                                          
  3. **Content quality gates**: Sentiment and Readability as deterministic thresholds before expensive LLM evaluation — e.g., skip `Correctness()` if
  output sentiment is negative or readability grade exceeds the target audience.

  4. **Regression detection**: LevenshteinRatio detects output drift between model versions — fast and deterministic.                                     
   
  > #### Why is this use case valuable to support for MLflow users in general?                                                                            
                  
  - Readability fills a documented migration gap from 3.4.0
  - BERTScore and METEOR are the two most-cited NLP evaluation metrics after BLEU/ROUGE (which MLflow covers via RAGAS)
  - All 5 metrics are free, fast, and reproducible — no LLM calls                                                                                         
  - These are standard NLP evaluation metrics available in nltk, HuggingFace evaluate, and scikit-learn — widely adopted across academia and industry.
  MLflow's third-party integrations (RAGAS, DeepEval) focus on LLM-judged and RAG-specific metrics, leaving these foundational NLP metrics without a      
  native scorer equivalent.
                                                                                                                                                          
  > #### Why is it currently difficult to achieve this use case?

  Users must install external packages, write custom `@scorer` functions, handle edge cases (empty text, encoding), and format results as `Feedback`      
  objects. For readability specifically, the deprecated `mlflow.metrics` functions no longer work with `mlflow.genai.evaluate()`.
                                                                                                                                                          
  ### Details     

  Implementation extends `Scorer` base class directly — no dependency on other proposals. Each scorer:

  - Returns `Feedback` with a continuous score (0.0–1.0) and rationale string                                                                             
  - Uses `source_type="CODE"` and `source_id="mlflow.scorers.<ClassName>"`
  - Handles edge cases (empty strings, None values) gracefully with `Feedback(error=e)`                                                                   
  - Lazy-imports optional dependencies at instantiation, not at module load                                                                               
   
  Dependencies:                                                                                                                                           
  - `Readability` → `textstat` (same dependency the deprecated legacy metric used)
  - `BERTScore` → `bert-score`
  - `METEOR` → `nltk`
  - `Sentiment` → `nltk` (VADER)                                                                                                                          
  - `LevenshteinRatio` → `python-Levenshtein` with `difflib.SequenceMatcher` fallback (zero-dep)
                                                                                                                                                          
  ### Files to add

  | File | Content |                                                                                                                                      
  |---|---|
  | `mlflow/genai/scorers/builtin_nlp_scorers.py` | 5 scorer implementations |                                                                            
  | `tests/genai/scorers/test_builtin_nlp_scorers.py` | Unit tests for all scorers |
                                                                                                                                                          
  Plus updates to `mlflow/genai/scorers/__init__.py` for exports.
                                                                                                                                                          
  ### What machine learning domain(s) is this feature request about?

  - [x] `domain/genai`: LLMs, Agents, and other GenAI-related use cases                                                                                   
   
  ### What area(s) of MLflow is this feature request about?                                                                                               
                  
  - [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

### PR DESCRIPTION
**Pull request by SahilKumar75**
### Related Issues/PRs

Closes #22424

### What changes are proposed in this pull request?

This PR implements 5 deterministic NLP scorers for fast, reproducible text evaluation without requiring LLM API calls:

- **BERTScore** - Semantic similarity using BERT embeddings
- **METEOR** - Machine translation evaluation metric  
- **Readability** - Text complexity metrics (Flesch-Kincaid, etc.)
- **Sentiment** - Sentiment analysis using VADER
- **LevenshteinRatio** - Edit distance similarity

These scorers provide objective metrics ideal for LLM fine-tuning evaluation, content quality monitoring, and offline evaluation workflows.

**Usage example:**
```python
from mlflow.genai.scorers import BERTScore, Readability, Sentiment

bert_scorer = BERTScore()
feedback = bert_scorer(
    outputs="The cat sat on the mat",
    expectations={"expected_output": "A cat was sitting on a mat"}
)
```

**Installation:**
```bash
pip install mlflow[nlp-scorers]
```

**Files changed:**
- Added: `mlflow/genai/scorers/deterministic_nlp_scorers.py`
- Added: `tests/genai/scorers/test_deterministic_nlp_scorers.py`
- Modified: `mlflow/genai/scorers/__init__.py`
- Modified: `pyproject.toml`

### How is this PR tested?

- [x] New unit/integration tests

Comprehensive test suite with 40+ test cases covering:
- Basic functionality for each scorer
- Custom parameter configurations
- Error handling (missing dependencies, invalid inputs)
- Edge cases (identical texts, empty strings)
- Integration scenarios

Run tests:
```bash
pytest tests/genai/scorers/test_deterministic_nlp_scorers.py -v
```

### Does this PR require documentation update?

- [x] Yes. I've updated:
- [x] API references (docstrings in implementation)

### Does this PR require updating the MLflow Skills repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added 5 built-in deterministic NLP scorers (BERTScore, METEOR, Readability, Sentiment, LevenshteinRatio) for fast, reproducible text evaluation without LLM API calls. These scorers are ideal for LLM fine-tuning evaluation and offline workflows.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Comment by SahilKumar75** [comment]
This PR has been superseded by #23507 which includes the complete PR template and proper DCO sign-off. Please refer to #23507 for the latest implementation.
